torchdata/datapipes/iter/util/prefetcher.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import threading
+import time
+
+from typing import Optional
+
+from torchdata.dataloader2 import communication
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+PRODUCER_SLEEP_INTERVAL = 0.0001  # Interval between buffer fullfilment checks
+CONSUMER_SLEEP_INTERVAL = 0.0001  # Interval between checking items availablitity in buffer
+
+
+class _PrefetchData:
+    def __init__(self, source_datapipe, buffer_size):
+        self.run_prefetcher = True
```

Comment:
And, should we add a thread lock around `run_prefetcher`?

Reply:
my thread lock is GIL =)

Reply:
Lol. Kind make sense.